rdm/0013-access-levels.md

```diff
@@ -0,0 +1,173 @@
+| admin                                  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
+
+
+**Proposed schema fields**
+
+```json
+{
+  "definitions": {
+    "permission_identity": {
+      "type": "object",
+      "description": "Identity that can be targeted by a permission.",
+      "additionalProperties": false,
+      "type": "object",
+      "properties": {
+        "id": {
+          "type": "string",
+          "description": "Per type unique id.",
+        },
+        "scheme": {
+          "type": "string",
+          "description": "Kind of permission_identity. What the id represents.",
+          "enum": [
+            "person",
+            "role",
+            "org"
+          ]
+        }
+      },
+      "required": ["id", "scheme"]
+    }
+  },
+  "access_levels": {
+    "type": "object",
+    "description": "Assigned access levels to permission identities",
+    "properties": {
+      "metadata_reader": {
+        "type": "array",
+        "items": { "$ref": "#/definitions/permission_identity" }
+      },
+      "metadata_curator": {
+        "type": "array",
+        "items": { "$ref": "#/definitions/permission_identity" }
+      },
+      "files_reader": {
+        "type": "array",
+        "items": { "$ref": "#/definitions/permission_identity" }
+      },
+      "files_curator": {
+        "type": "array",
+        "items": { "$ref": "#/definitions/permission_identity" }
+      },
+      "admin": {
+        "type": "array",
+        "items": { "$ref": "#/definitions/permission_identity" }
+      }
+    },
+    "additionalProperties": true
+  }
+}
+
+```
+
+The `access_levels` field should be part of the system (`internal`) metadata
+schema section.
+
+**PermissionPolicy Configuration + Generator**
+
+InvenioRDM's default `PolicyPermission` will have the `AllowedByAccessLevel`
+generators set by default:
+
+```python
+class RecordPermissionPolicy(BasePermissionPolicy):
+    """Access control configuration for records."""
+    # Other permissions omitted for clarity but same pattern applies
+    can_read = [..., AllowedByAccessLevel('read')]
+```
+
+This Generator encapsulates the logic to select the action, verify the access
+level and allow the permitted identities.
+
+Access levels are a means to add permissions, not subtract them.
+
+
+## How we teach this
+
+Access levels map well to every-day permission granting. Bob wants Alice to
+check his record's metadata before making it fully public: he gives her the
+Metadata Curator access level on his record. What needs to be clear is what
+each access level means, so that the right one can be picked. The UI should
+provide that information when it comes time to choose.
+
+
+## Drawbacks
+
+Because InvenioRDM decides the mapping of access level to permissions, there
+will always be administrators who have different needs. They will have to
+code their own, but the structure allows it.
```

Comment:
The use of a policy allows customization code wise but not naming wise (e.g. `files_curator`), this would require a change in the schema.

Would it be possible to create this levels as a dynamically loaded abstraction (i.e. roles?) that then maps the provides to needs of the datamodel (`owner`, `files_restricted`, etc.).

Nonetheless, for explicit (i.e. sharing link) we do need something like this.

Reply:
The implementation here: https://github.com/inveniosoftware/invenio-records-permissions/pull/29 might shed some light on the dynamic possibilities. Essentially, as long as the access level (whatever name it has) is slotted for the right actions, those will be used in the search without further schema changes.

That being said having greater control over these access levels could be done via a role / database association system... but I would wait for that since it is heavy. Maybe some configuration could go a long way (even then, I am not a big fan of kicking configuration down to the instance administrator).

Reply:
> but I would wait for that since it is heavy

Agree

> even then, I am not a big fan of kicking configuration down to the instance administrator

Also agree.

This was mostly a comment. Let's see what @slint @lnielsen think about it.

Reply:
I think I see what you mean.

The missing piece I haven't put in there, but was going to be added eventually to address that is the mapping from: human-readable access level (e.g.,  `"Metadata Curator"`) to this internal access level key name (`"metadata_curator"`). Then the two can change independently as long as they are connected with each other. I could definitely see `"Metadata Curator"` instead be `"Edit Record Metadata"`. This way we can keep `metadata_curator` internally without the UI affecting the internals.